### PR TITLE
feat: add toggle for distributed backend training framework

### PR DESCRIPTION
**Pull request by jaideepr97**
Adds an option and flag for picking a distributed training backend framework between FSDP and DeepSpeed

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the
  [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [x] [Changelog](https://github.com/instructlab/instructlab/blob/main/CHANGELOG.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Functional tests have been added, if necessary.
- [ ] E2E Workflow tests have been added, if necessary.

### CODE REVIEW

File: src/instructlab/clickext.py [hunk 1]
```diff
@@ -197,6 +198,10 @@ def consume_value(
         value, source = super().consume_value(ctx, opts)
         # fix parameter source for config section that are mis-reported
         # as DEFAULT source instead of DEFAULT_MAP source.
+
+        if isinstance(value, enum.Enum):
```
Comment (cdoern) [inline src/instructlab/clickext.py:202]: can we maybe comment why we are doing this? I know you know and I know, but just for others!

File: src/instructlab/model/train.py [hunk 2]
```diff
@@ -488,6 +494,15 @@ def get_files(directory: str, pattern: str) -> list[str]:
                 f"Data path must be to a valid .jsonl file. Value given: {data_path}"
             )
 
+        if distributed_training_backend:
```
Comment (cdoern) [inline src/instructlab/model/train.py:497]:
```suggestion
        if distributed_training_backend and  distributed_training_backend  not in DistributedBackend._value2member_map_:
```
?

File: tests/testdata/default_config.yaml [hunk 3]
```diff
@@ -264,6 +264,9 @@ train:
   # Allow CPU offload for deepspeed optimizer.
   # Default: False
   deepspeed_cpu_offload_optimizer: false
+  # Pick a distributed training backend framework for GPU accelerated full fine-tuning.
+  # Default: fsdp
+  distributed_training_backend: fsdp
```
Comment (cdoern) [inline tests/testdata/default_config.yaml:269]: fsdp as the default? is this the right behavior? I know its what we want to move towards, but not sure if its stable enough yet.

File: src/instructlab/clickext.py (hunk 1 above)
Comment (jaideepr97) [reply src/instructlab/clickext.py:207]: added

File: src/instructlab/model/train.py (hunk 2 above)
Comment (jaideepr97) [reply src/instructlab/model/train.py]: fixed, ty

File: tests/testdata/default_config.yaml (hunk 3 above)
Comment (jaideepr97) [reply tests/testdata/default_config.yaml]: i thought about it, seemed like a toss-up either way, looking at the future use cases
changed it back to deepspeed now

**Comment by cdoern** [comment]
`tox -e fix` or `make fix` should alleviate your ruff issues. try doing this stuff locally before pushing if possible

**Review by cdoern — APPROVED** [review]

**Review by leseb — APPROVED** [review]

**Comment by cdoern** [comment]
e2e CI passed here: https://github.com/instructlab/instructlab/actions/runs/11109142248 manualy merging as mergify is broken

**Pull request merged**